### PR TITLE
Batched uri resolution

### PR DESCRIPTION
**Pull request by jackrobison**
Goes with https://github.com/lbryio/lbry/pull/652 and https://github.com/lbryio/lbryum/pull/103

### CODE REVIEW

File: build/build.sh [hunk 1]
```diff
@@ -32,6 +32,9 @@ if [ "$FULL_BUILD" == "true" ]; then
   python "$BUILD_DIR/set_build.py"
 fi
 
+pip install git+https://github.com/lbryio/lbryschema.git
```
Comment (lyoshenka) [inline build/build.sh]: why is this here? requirements should be in requirements.txt

File: lbrynet/core/Wallet.py [hunk 2]
```diff
@@ -702,9 +702,13 @@ def _handle_claim_result(self, results, update_caches=True):
 
         if 'error' in results:
             if results['error'] in ['name is not claimed', 'claim not found']:
-                raise UnknownNameError(results['error'])
-            else:
-                raise Exception(results['error'])
+                if 'claim_id' in results:
```
Comment (lyoshenka) [inline lbrynet/core/Wallet.py]: this feels wrong

File: lbrynet/core/Wallet.py [hunk 3]
```diff
@@ -798,27 +802,35 @@ def _handle_claim_result(self, results, update_caches=True):
...
-        if resolve_results and 'claim' in resolve_results:
-            claim_id = resolve_results['claim']['claim_id']
-        certificate_id = None
-        if resolve_results and 'certificate' in resolve_results:
-            certificate_id = resolve_results['certificate']['claim_id']
-
-        result = yield self._handle_claim_result(resolve_results, cached_claim is None)
-        if claim_id:
-            yield self._storage.save_claim_to_uri_cache(uri, claim_id, certificate_id)
+    def resolve(self, check_cache, page, page_size, *uris):
```
Comment (lyoshenka) [inline lbrynet/core/Wallet.py]: its weird that `*uris` (the thing you care about most of the time) comes last here. it should be the first arg, and the rest should come after and have sane default values. so, something like this:

```
resolve(uris, page=1, page_size=20, check_cache=True)
```

Also, I know arrays are 0-indexed, but "page 0" makes no sense. Pagination starts on page 1.

File: lbrynet/lbrynet_daemon/Daemon.py [hunk 4]
```diff
@@ -787,9 +787,10 @@ def get_est_cost_using_known_size(self, uri, size):
 
         cost = self._get_est_cost_from_stream_size(size)
 
-        resolved = yield self.session.wallet.resolve_uri(uri)
-        if 'claim' in resolved:
-            claim = ClaimDict.load_dict(resolved['claim']['value'])
+        resolved = yield self.session.wallet.resolve(True, 0, 10, uri)
```
Comment (lyoshenka) [inline lbrynet/lbrynet_daemon/Daemon.py]: Please use named params here in `resolve(True, 0, 10, uri)`. `True`, `0`, and `10` are meaningless when you read this code. Or if you rearrange params as in the above comment, you only have to pass in the uris.

File: lbrynet/lbrynet_daemon/Daemon.py [hunk 5]
```diff
@@ -1459,19 +1461,22 @@ def jsonrpc_claim_show(self, name=None, txid=None, nout=None, claim_id=None):
...
     @defer.inlineCallbacks
     @AuthJSONRPCServer.flags(force='-f')
-    def jsonrpc_resolve(self, uri, force=False):
+    def jsonrpc_resolve(self, force=False, uri=None, uris=[]):
         """
         Resolve a LBRY URI
 
         Usage:
-            resolve <uri> [-f]
+            resolve (<uri> | --uri=<uri>) [<uris>...] [-f]
```
Comment (lyoshenka) [inline lbrynet/lbrynet_daemon/Daemon.py]: this is weird. shouldn't it just be `resolve [-f] uris...`? What's the point of `--uri` here? Is this just to maintain BC? If so, its probably best to fix this by handling this in the function itself (e.g. if you get one uri and its of the form `uri=whatever`, then remove the `uri=`). This way people are not encouraged to use the bad usage.

Comment (lyoshenka) [inline lbrynet/lbrynet_daemon/Daemon.py]: Then you also dont need the weird special case of one uri being resolved. You always return a dict of resolved uris, keyed by uri (though this does break BC and I don't see a clean fix for it off the top of my head).

File: lbrynet/lbrynet_daemon/Daemon.py [hunk 6]
```diff
@@ -1587,9 +1573,13 @@ def jsonrpc_get(self, uri, file_name=None, timeout=None, download_directory=None
         timeout = timeout if timeout is not None else self.download_timeout
         download_directory = download_directory or self.download_directory
 
-        resolved = yield self.session.wallet.resolve_uri(uri)
+        resolved_result = yield self.session.wallet.resolve(True, 0, 10, uri)
```
Comment (lyoshenka) [inline lbrynet/lbrynet_daemon/Daemon.py]: another place where the `page` and `page_size` could be avoided if they are made defaults. DRY, maaaaan :-)

File: lbrynet/lbrynet_daemon/Daemon.py [hunk 7]
```diff
@@ -2165,6 +2155,81 @@ def jsonrpc_claim_list(self, name):
         claims = yield self.session.wallet.get_claims_for_name(name)
         defer.returnValue(claims)
 
+    @AuthJSONRPCServer.auth_required
+    @defer.inlineCallbacks
+    def jsonrpc_claims_in_channel_list(self, page=0, page_size=10, uri=None, uris=[]):
```
Comment (lyoshenka) [inline lbrynet/lbrynet_daemon/Daemon.py]: we talked about this being in `claim_list`. we should not have a `claims_in_channels_list` RPC command. its dirty.

File: tests/unit/lbrynet_daemon/test_Daemon.py [hunk 8]
```diff
@@ -53,7 +54,7 @@ def get_test_daemon(data_rate=None, generous=True, with_fee=False):
             {"fee": {"USD": {"address": "bQ6BGboPV2SpTMEP7wLNiAcnsZiH8ye6eA", "amount": 0.75}}})
     daemon._resolve_name = lambda _: defer.succeed(metadata)
     migrated = smart_decode(json.dumps(metadata))
-    daemon.session.wallet.resolve_uri = lambda _: defer.succeed({'claim': {'value': migrated.claim_dict}})
+    daemon.session.wallet.resolve = lambda *_: defer.succeed({"test": {'claim': {'value': migrated.claim_dict}}})
```
Comment (lyoshenka) [inline tests/unit/lbrynet_daemon/test_Daemon.py]: shouldn't we test resolving multiple uris too?

File: build/build.sh (hunk 1 above)
Comment (jackrobison) [inline build/build.sh]: it's because when `pip install git+...`'ing, pip only looks at setup.py, so if the thing you're installing also depends on a non-pypy package it needs to be installed first.

Comment (jackrobison) [inline build/build.sh]: actually, since build.sh is just running `pip install -r requirements.txt` to install lbrynet, the lbryum branch should indeed be specified in requirements.txt

**Comment by kaykurokawa** [comment]
What should happen if a name is unclaimed? 

Before this PR, it was returning None.

Right now it is  returning dict {u'error': u'Name lbry://someunclaimedname1 is unknown'}

I think either way is ok. But others should state their opinion and behavior should be clarified in the docstring.

**Comment by lyoshenka** [comment]
As per the tech meeting: `[u'error': u'Name lbry://someunclaimedname1 is unknown'}` is better, Jack should update docstring.

File: lbrynet/core/Wallet.py (hunk 2 above)
Comment (jackrobison) [inline lbrynet/core/Wallet.py]: Added UnknownClaimID and UnknownURI exceptions.

File: lbrynet/lbrynet_daemon/Daemon.py (hunk 5 above)
Comment (jackrobison) [inline lbrynet/lbrynet_daemon/Daemon.py]: yeah, [-f] should be at the beginning.

BC aside, it's (<uri>) [<uris>...] also so that at least one uri is required

File: tests/unit/lbrynet_daemon/test_Daemon.py (hunk 8 above)
Comment (jackrobison) [inline tests/unit/lbrynet_daemon/test_Daemon.py]: Yes - we should be testing Wallet.resolve in general. This test is just doing get_est_cost, which only does one resolution
